### DISCUSSION TITLE
Google Drive Synchronization

### DISCUSSION THREAD

**Discussion by rumplestilzken** [vkbo/novelWriter#2009] [Ideas]
We discussed briefly the possibility of Google Drive synchronization in an issue I posted about getting novelWriter working on a Chromebook #1676 

Are you still interested in that collaboration? I would very much like to implement this feature if you are on board. A set of guidelines for how you would like things to work would be a good start so i'm not going totally AWOL. But I'm willing to do the grunt work. rclone can do synchronization with Google Drive, i'm not sure if you would want it to work like it does or if you would just like it to work with the filesystem that gets created when you mount a Google Drive using rclone or similar software. Both have their pros and cons.

**Comment by vkbo** [comment]
Hi!

Depends on what it entails. I don't want to add features that interact with APIs. At least not at this point. But if there is a way to interact with the file system or mount that works with Google drive, it's maybe worth considering. Especially if it opens up a new platform.

File I/O is handled in multiple places in novelWriter, but most of it is passed through a small set of core classes, so if we're talking about adding a layer of abstraction, that should be fine. There's already some abstraction for storage, but it handles it on the file path level, not the I/O stream itself.

**Reply by rumplestilzken** [reply]
Got it.  I will start working on a prototype that works with rclone in mind, which should in theory open up a potential for 20+ cloud formats. If we need to tweak it to be more compatible with your codebase or mindset we can worry about that once it is working.

Is I/O Streams something you had in mind moving forward or are local paths something you'd like to keep?

**Reply by vkbo** [reply]
But what does "works with rclone" mean in this context?

**Reply by vkbo** [reply]
The storage class is written to handle local paths. Or anything Python's pathlib can handle anyway. I'd like to keep it like that. The point of the storage class is to allow for handling zip files and temporary paths seamlessly to the rest of the code. The plan is to allow the project to be stored in an archived file format, like Open Document and MS Office does.

Anything more complex would probably be best served with a custom context manager that handles I/O at an even lower level than pathlib paths. But that requires a more extensive rewrite of a number of core classes.

**Reply by rumplestilzken** [reply]
rclone is a utility that manages mounted drives on windows, linux and OSX. It is how i sync my google drive to my desktop. It mounts to an empty folder, where I am able to drag and drop files from my desktop to and from my google drive. It supports a large number of cloud solutions. When used with a caching parameter, it allows you to read as if it were local and write a file locally in a caching location, sync to the cloud and then remove the cached files.

We would need to determine if the location that the project is saved is a local folder or a mounted remote drive. 

The way it would need to work, we would manage a cached copy of a NWDocument or other such item in a cache location, determined by rclone, while the document is active or the project files are being updated, which we would likely need to verify and possibly remount the drive with, for certain expectations to be met, for instance virtual file system writes being cached if the remote drive was not mounted with this parameter, which would be abstracted as far as needed to take all the rclone-ness out of key classes that would work for other mounting platforms/applications. In the case that we would need to change mount parameters and remount or just notify the user that the mount is not is proper configuration to be used and must be manually remounted we would need to handle at least one of these situations.

There are things we would need to worry about, such as keeping a NWDocument or other object "Alive" while active and in use within the editor, there for in the cached location and not the mounted location. We would need to either support merging updates from the mounted drive while files are active or stop syncing the files to perform a one way sync from the latest "active" document/location when it is no longer "active".

We could add a UI element under "Project" for projects on mounted drives to manually sync instead of having a live auto save based sync, bringing the whole project into the cache location until the application closes and then allowing the syncing application (rclone) to sync all files that have changed back to the remote drive.

I've got a page of notes so far on how it theoretically could work. I have reviewed some of the code to get my grips on what is already there and how things are supported. I have built the release, so i know i'm good to go on starting to make changes, I will continue to review your code to get a handle on trying to keep things according to conventions already set in the codebase.

Thoughts?

**Reply by rumplestilzken** [reply]
> The storage class is written to handle local paths. Or anything Python's pathlib can handle anyway. I'd like to keep it like that. The point of the storage class is to allow for handling zip files and temporary paths seamlessly to the rest of the code. The plan is to allow the project to be stored in an archived file format, like Open Document and MS Office does.
> 
> Anything more complex would probably be best served with a custom context manager that handles I/O at an even lower level than pathlib paths. But that requires a more extensive rewrite of a number of core classes.

File paths will work just fine for what I'm thinking of, no worries. Just checking.

Having an archive would make some of what i've already said change slightly, but nothing too big. If that is the way the project is heading, i will keep that in mind.

**Reply by vkbo** [reply]
I'm planning to create a temporary project location where files are read/written when the project is open. That's why the storage class has two paths. One for the where it is stored when closed, and the temporary path where it is when open. This is designed to accommodate zipped projects. Currently, when the project is open from a folder, those two paths are identical.

I think if we treat your case in a similar manner, then that would work fine too. It's just about setting a flag and adding a handler for making the final write at project close. And for when auto-save triggers.

If this is the solution, then we need to look a little closer on what I'm planning for the zip feature. I have a branch on it, but it's way out of date with current main, but the storage class isn't very different. You can have a look at: https://github.com/vkbo/novelWriter/blob/new_single_file/novelwriter/core/storage.py

**Reply by vkbo** [reply]
See for instance here:

```python

class NWStorage:
    """Core: Project Storage Class

    The class that handles all paths related to the project storage.
    """

    MODE_INACTIVE = 0
    MODE_INPLACE  = 1
    MODE_ARCHIVE  = 2

    def __init__(self, project: NWProject) -> None:
        self._project = project
        self._storagePath = None
        self._runtimePath = None
        self._lockFilePath = None
        self._lockedBy = None
        self._openMode = self.MODE_INACTIVE
        self._readOnly = True
        self._ready = False
        self._error = NWStorageError.NO_ERROR
        self._exception = None
        return
```

The `self._runtimePath` variable is the one used when opening, while the `self._storagePath` is used on extract/compress. I guess we can inject similar logic in the relevant methods for sync caching. Or at least use that as a starting point. Which mode the class operates on is determined by the `MODE_` constants.

The `self._lockFilePath` should explain itself, and is there to prevent multiple instances writing to the same location. `self._readOnly` is meant to be used when opening a project from backup. I haven't entirely decided how to deal with that yet though.

**Reply by vkbo** [reply]
I haven't decided how to cache open projects yet. I'm open to ideas.

My initial idea was to use the data location, which is `~/.local/share/novelwriter/`, create a `projects` folder there, and for each project, create a folder that is the sha1 hash of the absolute storage path. This is the simplest way to associate the cache folder directly with the corresponding storage location so that we can detect if the project was not properly closed last time (in which case the sha hash folder will already exist, and we can trigger a recovery process).

**Reply by rumplestilzken** [reply]
> I haven't decided how to cache open projects yet. I'm open to ideas.
> 
> My initial idea was to use the data location, which is `~/.local/share/novelwriter/`, create a `projects` folder there, and for each project, create a folder that is the sha1 hash of the absolute storage path. This is the simplest way to associate the cache folder directly with the corresponding storage location so that we can detect if the project was not properly closed last time (in which case the sha hash folder will already exist, and we can trigger a recovery process).

By placing the cached open projects in the same folder as the projects exist, but in a hidden folder, you gain the possibility of syncing the unfinished files to the cloud as well. The caching and syncing will automatically be handled by the sync provider(rclone here). Then, when closing the application, or on auto save, we rebuild the archive, the sync provider sees that the file has changed, moves it to its own cache location, syncs, and deletes the leftovers. Then we remove the hidden application cached project as long as we know nothing else has been modified since last archive and it all just kinda fits together. We can leverage the syncing remote caching system to make sure our local application project cache is always synced to the cloud for maximum recovery potential as well, until archiving and then archive and clean up.

I can work on a simple archiving process that can be swapped out to follow this methodology and give you something to stand on and you can do what you'd like with it if that interests you. Just build a small framework that would handle the actual archiving on save and autosave in a way that will allow you to expand or change how you see fit, after i have reviewed the code you mentioned earlier.

**Reply by rumplestilzken** [reply]
In addition to the loading of the new archive if it presents itself. I don't think putting something small together would be too much of an effort and i generally aim for maintenance and reusability/configurability so you wouldn't have to worry about not being able to plop your own solution in when the time comes.

**Reply by vkbo** [reply]
Extracting the archive next to itself is also an idea I've considered. I may end up doing that in the end, for the sake of transparency. The principle of novelWriter is to ensure that text is saved frequently and safely. I wouldn't necessarily have the folder hidden even. Just make sure the name makes it clear it is temporary. If not for this, I could easily just cache everything in memory since it isn't much data even for a large project.

**Reply by vkbo** [reply]
I may start working on this solution soon actually. I'm just taking a break from developing new features right now as I've spent a lot of time on the last two releases, and I want to do other things for a bit.

**Comment by rumplestilzken** [comment]
It just occurred to me that if you package it all up, you then have to keep track of revision progress as well. 

I think if you want to implement the archived project, you should automatically have it backed up into a reposity. Git, svn, whatever. I think making the application archive every x amount of minutes, including the back end local repository would be ideal. And it presents another opportunity to read through your writing and quickly compare revisions. 

This appoach stops zip archives from being loaded into the SCM which is going to bloat. 

Also having access to a graphical interface for the projects history using SCM would help a lot. 

Tldr; SCM integration before zip container work. Then revisions. Then special support for each popular SCM that would need to be supprt. The containerize the given project.

**Comment by rumplestilzken** [comment]
@vkbo Hey, just checking back in on this.

**Reply by vkbo** [reply]
On what part? If you're following up on the existing discussion thread, please don't start another. If you want to start a new one, you need to be more specific.
